### PR TITLE
CMake: Switch to exported namespace targets and prefer config mode

### PR DESCRIPTION
**Pull request by Ryanf55**
This allows use of config mode in boost to find it. Regardless of whether boost is found via config or module, the alias targets can, and are preferred, to be used in CMake. 

Reference: https://cmake.org/cmake/help/latest/module/FindBoost.html#boost-cmake
Relates to https://github.com/ompl/ompl/issues/1065

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -48,9 +48,7 @@ set_package_properties(Boost PROPERTIES
     URL "https://www.boost.org"
     PURPOSE "Used throughout OMPL for data serialization, graphs, etc.")
 set(Boost_USE_MULTITHREADED ON)
-set(Boost_NO_BOOST_CMAKE ON)
-find_package(Boost 1.58 QUIET REQUIRED COMPONENTS serialization filesystem system program_options)
```
Comment (Ryanf55) [inline CMakeLists.txt:52]: I can keep QUIET if you want...

File: src/ompl/CMakeLists.txt [hunk 2]
```diff
@@ -36,15 +36,20 @@ source_group("OMPL Headers" FILES "${OMPL_HEADERS}")
 # build the library
 if(MSVC)
     add_library(ompl STATIC ${OMPL_SOURCE_CODE})
-else(MSVC)
```
Comment (Ryanf55) [inline src/ompl/CMakeLists.txt:39]: This is no longer recommended in CMake btw

File: src/ompl/CMakeLists.txt [hunk 3]
```diff
@@ -36,15 +36,20 @@ source_group("OMPL Headers" FILES "${OMPL_HEADERS}")
...
-endif(MSVC)
+endif()
 target_link_libraries(ompl
     ${OMPL_LINK_LIBRARIES}
     ${Boost_SERIALIZATION_LIBRARY}
     ${Boost_FILESYSTEM_LIBRARY}
     ${Boost_SYSTEM_LIBRARY}
-    ${CMAKE_THREAD_LIBS_INIT})
+    ${CMAKE_THREAD_LIBS_INIT}
+    Boost::filesystem
```
Comment (Ryanf55) [inline src/ompl/CMakeLists.txt:48]: Linking alias targets ensures they actually exist, vs variables which can silently fail.

File: CMakeLists.txt [hunk 4]
```diff
@@ -48,9 +48,7 @@ set_package_properties(Boost PROPERTIES
     URL "https://www.boost.org"
     PURPOSE "Used throughout OMPL for data serialization, graphs, etc.")
 set(Boost_USE_MULTITHREADED ON)
-set(Boost_NO_BOOST_CMAKE ON)
```
Comment (Ryanf55) [inline CMakeLists.txt:51]: I cannot tell what reason a user would enable this. CMake developers recommend the use of CONFIG mode, whenever possible, but this disables it.

Comment (mamoll) [reply CMakeLists.txt:51]: This was probably put in when the Boost CMake support was still experimental. I'm in favor of this change.

File: CMakeLists.txt [hunk 5]
```diff
@@ -48,9 +48,7 @@ set_package_properties(Boost PROPERTIES
...
-find_package(Boost 1.58 QUIET REQUIRED COMPONENTS serialization filesystem system program_options)
-include_directories(SYSTEM ${Boost_INCLUDE_DIR})
```
Comment (mamoll) [inline CMakeLists.txt:53]: Why is this line deleted? If I have set the environment variable `BOOST_ROOT` to some non-standard boost installation, will the correct include path be used?

**Review by mamoll — COMMENTED** [review]
I have one question, but I'm ok with this change.

File: CMakeLists.txt (hunk 5 above)
Comment (Ryanf55) [reply CMakeLists.txt:53]: The includes are brought in through the call to `target_link_libraries`. The exported namespace targets that are found above already have the properties set. 

According to the [docs](https://cmake.org/cmake/help/latest/module/FindBoost.html#hints), it looks like `BOOST_ROOT` is used by the module as a hint of where to find boost. Therefore, this would be my expected behavior if you have installed boost in a custom location.
1. Cmake tries to find it boostConfig.cmake from find_package and fails since it's not on cmake prefix path
2. You gave a hint for the module to find boost from `BOOST_ROOT`
3. The find succeeds, and populates the imported namespace targets
4. `target_link_libraries` functions the same.

If you have a system without boost installed in system headers and instead in a non-standard location, please try. 

If you think this is a common way people find boost, perhaps one could add it to CI?

Comment (mamoll) [reply CMakeLists.txt:53]: > The includes are brought in through the call to target_link_libraries. The exported namespace targets that are found above already have the properties set.

Neat! I learnt something today. I tried it and it works with boost in a non-standard location.

File: src/ompl/CMakeLists.txt (hunk 3 above)
Comment (mamoll) [reply src/ompl/CMakeLists.txt:48]: Don't you need to add Boost::serialization and delete ${Boost_SERIALIZATION_LIBRARY} and ${Boost_FILESYSTEM_LIBRARY}?

Comment (Ryanf55) [reply src/ompl/CMakeLists.txt:45]: `Boost::serialization` is already there, but yes, I forgot to remove the old properties. Great catch. 
While I'm at it, I'll sort them alphabetically. 

Fixed in a9c2ba50184365cec6599e011346fa6b30ce329f

**Comment by Ryanf55** [comment]
Much like here I try to get OMPL to consume the namespaced exported targets from boost, `ompl` doesn't install an export set. 
https://cmake.org/cmake/help/latest/guide/tutorial/Adding%20Export%20Configuration.html

I can either add that functionality to this PR, or do it in a follow up. Let me know which you prefer.

**Review by zkingston — APPROVED** [review]
LGTM, thanks for the change. Tried on my machine with defaults and with boost in a non-standard location.
